<a href="https://colab.research.google.com/github/DinaShaim/GB_ML_MathStat/blob/main/%D0%94%D0%BE%D0%BC%D0%B0%D1%88%D0%BD%D1%8F%D1%8F_%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%B0_%D0%9C%D0%B0%D1%82%D1%81%D1%82%D0%B0%D1%82_%D0%A3%D1%80%D0%BE%D0%BA_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Домашняя работа 7
### Многомерный статистический анализ. Линейная регрессия.

**Задание 1** 

Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks): zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110], ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]. Используя математические операции, посчитать коэффициенты линейной регрессии, приняв за X заработную плату (то есть, zp - признак), а за y - значения скорингового балла (то есть, ks - целевая переменная). Произвести расчет как с использованием intercept, так и без.

**Решение**



Обозначим за $x$  заработную плату, а за $y$  - значения скорингового балла.

Для начала проведем расчеты с использованием intercept, т.е. для уравнения вида $y = a + b*x$.


In [13]:
import numpy as np
n=10

In [14]:
x = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
x

array([ 35,  45, 190, 200,  40,  70,  54, 150, 120, 110])

In [15]:
y = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
y

array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

In [16]:
b = (np.mean(x*y) - np.mean(x)*np.mean(y))/(np.mean(x**2) - np.mean(x)**2)
b

2.620538882402765

In [17]:
a = np.mean(y) - b*np.mean(x)
a

444.1773573243596

Получили уравнение вида: 
$$y = 444.177 + 2.62 * x$$

Теперь проведем вычисления без intercept, т.е. для уравнения вида $y =  b*x$. Для этого воспользуемся матричным методом. Преобразуем массивы $x, y$ к необходимому виду:

In [18]:
x1 = x.reshape((n,1))
x1

array([[ 35],
       [ 45],
       [190],
       [200],
       [ 40],
       [ 70],
       [ 54],
       [150],
       [120],
       [110]])

In [19]:
y1 = y.reshape((n,1))
y1

array([[401],
       [574],
       [874],
       [919],
       [459],
       [739],
       [653],
       [902],
       [746],
       [832]])

In [20]:
b = np.dot(np.linalg.inv(np.dot(x1.T,x1)),x1.T @ y1)
b

array([[5.88982042]])

Получили уравнение вида: 
$$y = 5.89 * x$$

**Задание 2** 

Посчитать коэффициент линейной регрессии при заработной плате (zp), используя градиентный спуск (без intercept).

**Решение**

Для удобства создадим фукнцию mse, расчитывающию функцию потерь:

In [22]:
def mse (b1, y=y, x=x, n=10):
  return np.sum((b1*x-y)**2)/n

Зададим параметры градиентного метода:

In [23]:
b1 = 0.1
alpha = 1e-6
alpha

1e-06

Осуществим вычисления градиентным методом

In [30]:
for i in range (5000):
  b1 -=alpha * (2/n) * np.sum((b1*x-y)*x)
  if (i)%500==0:
    print('Iteration: {i},  b1 = {b1},  mse = {mse}'.format(i=i, b1=b1, mse=mse(b1)))

Iteration: 0,  b1 = 5.889820420132673,  mse = 56516.85841571943
Iteration: 500,  b1 = 5.889820420132673,  mse = 56516.85841571943
Iteration: 1000,  b1 = 5.889820420132673,  mse = 56516.85841571943
Iteration: 1500,  b1 = 5.889820420132673,  mse = 56516.85841571943
Iteration: 2000,  b1 = 5.889820420132673,  mse = 56516.85841571943
Iteration: 2500,  b1 = 5.889820420132673,  mse = 56516.85841571943
Iteration: 3000,  b1 = 5.889820420132673,  mse = 56516.85841571943
Iteration: 3500,  b1 = 5.889820420132673,  mse = 56516.85841571943
Iteration: 4000,  b1 = 5.889820420132673,  mse = 56516.85841571943
Iteration: 4500,  b1 = 5.889820420132673,  mse = 56516.85841571943


**Задание 3** 

В каких случаях для вычисления доверительных интервалов и проверки статистических гипотез используется таблица значений функции Лапласа, а в каких - таблица критических точек распределения Стьюдента?

**Решение**

В качестве критерия используется случайная величина t-критерий Стьюдента, если  для нормально распределенной генеральной совокупности не известно ни среднее квадратическое отклонение, ни дисперсия. 

При решении задач, требующих применения интегральной теоремы Лапласа, пользуются специальными таблицами значений функции Лапласа.

**Задание 4**

Произвести вычисления как в пункте 2, но с вычислением intercept. Учесть, что изменение коэффициентов должно производиться на каждом шаге одновременно (то есть изменение одного коэффициента не должно влиять на изменение другого во время одной итерации).

**Решение**

Для удобства создадим фукнцию mse, расчитывающию функцию потерь:

In [32]:
def mse1 (b0,b1, y=y, x=x, n=10):
  return np.sum((b0+b1*x-y)**2)/n

In [47]:
b0 = 0.1
b1 = 0.1
alpha = 4e-5

In [48]:
for i in range (1000000):
 b0 -= alpha * (2/n) * np.sum((b0 + b1*x - y))
 b1 -= alpha * (2/n) * np.sum((b0 + b1*x - y)*x)
 if (i)%100000==0:
   print('Iteration: {i},  b0 = {b0},  b1 = {b1},  mse = {mse}'.format(i=i, b0=b0, b1=b1, mse=mse1(b0,b1)))

Iteration: 0,  b0 = 0.1559728,  b1 = 6.4798579486464005,  mse = 61296.64334529305
Iteration: 100000,  b0 = 385.8231521007457,  b1 = 3.0500429169636174,  mse = 7334.197318381918
Iteration: 200000,  b0 = 436.5084025964089,  b1 = 2.67698463440808,  mse = 6485.332961698472
Iteration: 300000,  b0 = 443.1694973698982,  b1 = 2.6279570267345185,  mse = 6470.671869347142
Iteration: 400000,  b0 = 444.0449035905899,  b1 = 2.6215137806612328,  mse = 6470.418651471703
Iteration: 500000,  b0 = 444.15995015234995,  b1 = 2.620667004284981,  mse = 6470.414278039572
Iteration: 600000,  b0 = 444.1750696603721,  b1 = 2.6205557202794627,  mse = 6470.414202504192
Iteration: 700000,  b0 = 444.17705667778745,  b1 = 2.620541095249555,  mse = 6470.4142011995855
Iteration: 800000,  b0 = 444.1773178131555,  b1 = 2.6205391732167946,  mse = 6470.414201177052
Iteration: 900000,  b0 = 444.1773521317654,  b1 = 2.6205389206217786,  mse = 6470.414201176664
